// @antv/layout/lib/layout/layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ILayout = exports.Layouts = exports.Layout = void 0;
var grid_1 = require("./grid");
var random_1 = require("./random");
var gForce_1 = require("./gForce");
var force_1 = require("./force");
var circular_1 = require("./circular");
var dagre_1 = require("./dagre");
var radial_1 = require("./radial");
var registy_1 = require("../registy");
var concentric_1 = require("./concentric");
var mds_1 = require("./mds");
var fruchterman_1 = require("./fruchterman");
var fruchterman_2 = require("./gpu/fruchterman");
var gForce_2 = require("./gpu/gForce");
var comboForce_1 = require("./comboForce");
var Layout = /** @class */ (function () {
    function Layout(options) {
        var layoutClass = registy_1.getLayoutByName(options.type);
        this.layoutInstance = new layoutClass(options);
    }
    Layout.prototype.layout = function (data) {
        return this.layoutInstance.layout(data);
    };
    Layout.prototype.updateCfg = function (cfg) {
        this.layoutInstance.updateCfg(cfg);
    };
    Layout.prototype.init = function (data) {
        this.layoutInstance.init(data);
    };
    Layout.prototype.execute = function () {
        this.layoutInstance.execute();
    };
    Layout.prototype.getDefaultCfg = function () {
        return this.layoutInstance.getDefaultCfg();
    };
    Layout.prototype.destroy = function () {
        return this.layoutInstance.destroy();
    };
    return Layout;
}());
exports.Layout = Layout;
// FIXME
// FOR G6
exports.Layouts = new Proxy({}, {
    get: function (target, propKey) {
        return registy_1.getLayoutByName(propKey);
    },
    set: function (target, propKey, value) {
        registy_1.registerLayout(propKey, value);
        return true;
    }
});
var ILayout;
(function (ILayout) {
    registy_1.registerLayout('grid', grid_1.GridLayout);
    registy_1.registerLayout('random', random_1.RandomLayout);
    registy_1.registerLayout('force', force_1.ForceLayout);
    registy_1.registerLayout('circular', circular_1.CircularLayout);
    registy_1.registerLayout('dagre', dagre_1.DagreLayout);
    registy_1.registerLayout('radial', radial_1.RadialLayout);
    registy_1.registerLayout('concentric', concentric_1.ConcentricLayout);
    registy_1.registerLayout('mds', mds_1.MDSLayout);
    registy_1.registerLayout('fruchterman', fruchterman_1.FruchtermanLayout);
    registy_1.registerLayout('fruchterman-gpu', fruchterman_2.FruchtermanGPULayout);
    registy_1.registerLayout('gForce', gForce_1.GForceLayout);
    registy_1.registerLayout('gForce-gpu', gForce_2.GForceGPULayout);
    registy_1.registerLayout('comboForce', comboForce_1.ComboForceLayout);
})(ILayout = exports.ILayout || (exports.ILayout = {}));
